fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biomes/SpawnBiomesMixin.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.biomes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.fabricmc.fabric.impl.biomes.BiomeLists;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.source.BiomeSource;
+
+@Mixin(BiomeSource.class)
+public class SpawnBiomesMixin 
+{
+	@Shadow
+	@Final
+	private static List<Biome> SPAWN_BIOMES;
+	
+	@Inject(at = @At("HEAD"), cancellable = true, method = "getSpawnBiomes")
```

Comment:
On further review of the code, this Mixin is bad. This should be a `RETURN` injection, copying the return value and appending your custom spawn biomes to it. No need to even cancel.

Reply:
```java
@Inject(method = "getSpawnBiomes", at = @At("RETURN"), cancellable = true)
private void fabric$addSpawnBiomes(CallbackInfoReturnable<List<Biome>> cir) {
  // Use a Set when wrapping for distinct value comparisons during append
  Set<Biome> biomes = new LinkedHashSet<>(cir.getReturnValue());
  // Only cancel callback if the collection was mutated by this call
  if (biomes.addAll(BiomeLists.SPAWN_BIOMES)) {
    // Re-wrap collection with List for expected type
    cir.setReturnValue(new ArrayList<>(biomes));
  }
}
```

Reply:
I don't have much experience with mixin aside from head injection, so thanks

Reply:
> RiverAssociate is there to be available for custom river biomes as well
> BiomeAssociate is for other layers where we need a weighted randomness generator based on Minecraft's LayerRandomnesSource

What I meant is: hide the implementation, move it away from api/, provide factories that create interfaces instead, so we can do optimization/tweak changes later without breaking mod expectations

Reply:
Ok will try the code provided for this mixin now